### PR TITLE
Allow work with CATs on user-mode (issue #30)

### LINKED ISSUE #30: Allow work with CATs on user-mode
@pcantera commented on [Mon Jul 09 2018](https://github.com/ingrammicro/concerto/issues/83)

New _Cloud Application_ commands:
**cio cloud-applications templates \<CMD>**:

- **list** > Lists CATs
	GET	/v2/plugins/tosca/cats

- **show** -> Shows CAT
	GET	/v2/plugins/tosca/cats/:cat_id

- **upload** -> Uploads a CAT
	POST /v2/plugins/tosca/cats
	PUT file
	PUT /v2/plugins/tosca/cats/:cat_id/parse_metadata

- **delete** -> Deletes a CAT
	DELETE /v2/plugins/tosca/cats/:cat_id

**cio cloud-applications deployments \<CMD>**:
- **list** -> Lists deployments
	GET	/v2/labels (Namespace == "cat:deployment")

- **show** -> Shows deployment
	GET	/v2/plugins/tosca/deployments/:deployment_id

- **deploy** -> Deploys a CAT
	POST /v2/plugins/tosca/cats/:cat_id/deployment_tasks
	GET /v2/plugins/tosca/cats/:cat_id/deployment_tasks/:deployment_task_id

- **delete** -> Deletes a deployment
	DELETE /v2/plugins/tosca/deployments/:deployment_id

# Expected Behavior

# Current Behavior

# Possible Solution

# Steps to Reproduce (for bugs)

    1.
    2.
    3.
    4.

# Context and environment

### PR DESCRIPTION
**Pull request by pcantera**
# Description
New _Cloud Application_ commands:
imagen

**cio cloud-applications templates \<CMD>**:
imagen

- **list** > Lists CATs
	GET	/v2/plugins/tosca/cats

- **show** -> Shows CAT
	GET	/v2/plugins/tosca/cats/:cat_id

- **upload** -> Uploads a CAT
	POST /v2/plugins/tosca/cats
	PUT file
	PUT /v2/plugins/tosca/cats/:cat_id/parse_metadata

- **delete** -> Deletes a CAT
	DELETE /v2/plugins/tosca/cats/:cat_id

**cio cloud-applications deployments \<CMD>**:
imagen

- **list** -> Lists deployments
	GET	/v2/labels (Namespace == "cat:deployment")

- **show** -> Shows deployment
	GET	/v2/plugins/tosca/deployments/:deployment_id

- **deploy** -> Deploys a CAT
	POST /v2/plugins/tosca/cats/:cat_id/deployment_tasks
	GET /v2/plugins/tosca/cats/:cat_id/deployment_tasks/:deployment_task_id

- **delete** -> Deletes a deployment
	DELETE /v2/plugins/tosca/deployments/:deployment_id

# Related Issue
Closes #30 

# Motivation and Context

# How Has This Been Tested?

# Screenshots (if appropriate)

# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: api/types/applications.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	ID           string  `json:"id" header:"ID"`
+	Name         string  `json:"name" header:"NAME"`
+	Version      string  `json:"version" header:"VERSION"`
+	Global       bool    `json:"global" header:"GLOBAL"`
+	UploadURL    string  `json:"upload_url,omitempty" header:"UPLOAD_URL" show:"noshow,nolist"`
+	VendorID     string  `json:"vendor_id" header:"VENDOR_ID"`
+	Inputs       []Input `json:"inputs" header:"INPUTS" show:"nolist"`
+	IsMock       bool    `json:"is_mock" header:"IS_MOCK" show:"nolist"`
+	ResourceType string  `json:"resource_type" header:"RESOURCE_TYPE" show:"nolist"`
+	//LabelableFields
```
Comment (pbanos) [inline api/types/applications.go:14]: This comment is not necessary I guess

File: api/cloudapplication/deployments_api.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return err
+	}
+
+	if err = utils.CheckStandardStatus(status, data); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (pbanos) [inline api/cloudapplication/deployments_api.go:89]: This actually only starts the undeployment process, you should return the representation of the deployment in the response, which contains the value of the label set to undeploying

File: api/types/applications.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	ID           string `json:"id" header:"ID"`
+	Name         string `json:"name" header:"NAME"`
+	Namespace    string `json:"namespace" header:"NAMESPACE" show:"nolist"`
+	Value        string `json:"value" header:"VALUE"`
+	CatID        string `json:"cat_id" header:"CAT_ID" show:"nolist"`
+	ResourceType string `json:"resource_type" header:"RESOURCE_TYPE" show:"nolist"`
+}
+
+// tosca task
+type DeploymentTask struct {
```
Comment (pbanos) [inline api/types/applications.go:36]: You shoud probably call this CloudApplicationDeploymentTask or CADeploymentTask, to make it clearer its relationship to CloudApplicationDeployment

File: cmd/cloud_application_deployments_cmd.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+func CloudApplicationDeploymentDelete(c *cli.Context) error {
+	debugCmdFuncInfo(c)
+	svc, formatter := WireUpCloudApplicationDeployment(c)
+
+	checkRequiredFlags(c, []string{"id"}, formatter)
+	err := svc.DeleteDeployment(c.String("id"))
+	if err != nil {
+		formatter.PrintFatal("Couldn't delete cloud application deployment", err)
+	}
+	return nil
```
Comment (pbanos) [inline cmd/cloud_application_deployments_cmd.go:131]: You should probably wait here until the deployment disappears

File: main.go [hunk 5]
```diff
@@ -72,6 +73,12 @@ var clientCommands = []cli.Command{
 		Usage:       "Manages brownfield resources, allowing users to discover and import servers, VPCs, floating IPs and volumes from different cloud accounts into the system.",
 		Subcommands: append(clientbrownfield.SubCommands()),
 	},
+	{
+		Name:        "cloud-applications",
+		ShortName:   "ca",
+		Usage:       "Manages cloud application templates -CATs- and deployments, allowing users to import and export of infrastructures using TOSCA (Topology and Orchestration Specification for Cloud Applications)",
```
Comment (chris-m-h) [inline main.go:79]: Remove comment about imports/exports, as this is a separate world

File: cmd/cloud_application_deployments_cmd.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		}
+		log.Info("State: ", deploymentTask.State)
+
+		if deploymentTask.State != "pending" {
+			if err = formatter.PrintItem(*deploymentTask); err != nil {
+				formatter.PrintFatal("Couldn't print/format result", err)
+			}
+			break
+		}
+		time.Sleep(5 * time.Second)
```
Comment (pbanos) [inline cmd/cloud_application_deployments_cmd.go:115]: This should be configurable through a CLI switch, and probably default to something longer such as 30 seconds

File: cmd/cloud_application_deployments_cmd.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		log.Info("State: ", deployment.Value)
+
+		// stops if something fails while undeploying
+		if deployment.Value != "undeploying" {
+			if err = formatter.PrintItem(*deployment); err != nil {
+				formatter.PrintFatal("Couldn't print/format result", err)
+			}
+			break
+		}
+		time.Sleep(10 * time.Second)
```
Comment (pbanos) [inline cmd/cloud_application_deployments_cmd.go:154]: Same as above

File: cmd/cloud_application_deployments_cmd.go (hunk 6 above)
Comment (pcantera) [reply cmd/cloud_application_deployments_cmd.go]: Done, implemented with a new optional input parameter

File: cmd/cloud_application_deployments_cmd.go (hunk 7 above)
Comment (pcantera) [reply cmd/cloud_application_deployments_cmd.go]: Done, implemented with a new optional input parameter

**Pull request merged**